src/tool/hpcrun/main.c

```diff
@@ -1264,14 +1253,13 @@ monitor_init_thread(int tid, void* data)
     hpcrun_thread_suppress_sample = true;
   }
 
-  hpcrun_safe_enter();
-
   TMSG(THREAD,"init thread %d",tid);
   void* thread_data = hpcrun_thread_init(tid, (local_thread_data_t*) data, ! hpcrun_thread_suppress_sample);
   TMSG(THREAD,"back from init thread %d",tid);
 
   hpcrun_threadmgr_thread_new();
 
+  // The new thread is initialized as "inside hpcrun", so we have to switch back
```

Comment:
The issue is that `hpcrun_safe_enter()` returns false if the `thread_data_t` isn't available yet, so any safety nests above this function will always be lost. This (unconditional) call is needed to make sure we are "outside hpcrun" before the application fires up and we start getting real data.

The alternative is to make the `inside_hpcrun` flag a thread-local variable instead of binding it to the `td`, that way it would always work (presuming we don't jump in before our static TLS initialization... which I'm not yet confident is impossible).